File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        int l = 0, r = nums.size()-1;
        int res = nums[r];
        while(l<=r){
            int m = (l+(r-l)/2);
            if(nums[l] <= nums[r]){
                res = min(res,nums[l]);
                break;
            }
            if(nums[m]>=nums[l]){
                l=m+1;
                res = min(res,nums[l]);
            }else{
                r=m-1;
                res = min(res,nums[m]);
            }
            cout << res << " \n";
        }
        return res;
    }
};

File: Data Structures & Algorithms/permutations/submission-2.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        n = len(nums)
        used = [False] * n
        curr = []

        def dfs():
            if(len(curr) == n):
                res.append(curr[:])
                return
            
            for i in range(n):
                if used[i]:
                    continue
                
                used[i] = True
                curr.append(nums[i])
                dfs()

                used[i] = False
                curr.pop()

        dfs()
        
        return res
            
        
